### PR TITLE
Jump types for yul functions.

### LINKED ISSUE #8839: Request: Jump markings on Yul functions
## Abstract

I'd like it if Yul functions had jump markings like Solidity functions do.  This would allow debuggers to distinguish between multiple invocations of the same function.  For Truffle Debugger it would allow me to distinguish local variables from one instance of a function, vs those same variables in a recursive invocation of that function.

## Motivation

While Truffle Debugger doesn't currently support assembly variables, [I have a PR open](https://github.com/trufflesuite/truffle/pull/3018) to make it do so.  However, it has no way to distinguish between local variables in recursive invocations of the same function, which will lead to misidentifying variables if a function is invoked recursively.

There is maybe one argument against this, which is that maybe jump markings should be reserved for Solidity stackframes and not Yul stackframes.  But I am doubtful that this is really a serious problem compared to that of not being able to distinguish Yul stackframes.

## Specification

Jumps into Yul functions should be marked with a jump in, and their returns should be marked with a jump out.

## Backwards Compatibility

I'm unclear on how breaking this change would be.  It could cause problems for anything that assumes that jump markings are *only* for Solidity stackframes and definitely not for Yul stackframes.  I don't know if that's a serious problem.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #8839

### CODE REVIEW

**Comment by chriseth** [comment]
@yann300 @haltman-at @alcuadrado @iamdefinitelyahuman 

This adds the "jump into" / "jump out of" tags also to `jump` / `jumpi` opcodes coming from yul functions, so this affects the ABI coder, inline assembly and so on. Can you please check if this breaks anything for your framework and / or is useful for you?

**Comment by chriseth** [comment]
Linux binary (might need specific libraries): https://365135-40892817-gh.circle-artifacts.com/0/solc
emscripten binary: https://365130-40892817-gh.circle-artifacts.com/0/soljson.js

**Comment by alcuadrado** [comment]
We still have to adapt our test suite to 0.6.9, so I can't be completely sure about it, but I think this will need some changes in Buidler. We track jump into and returns, to create something similar to a callstack. This change will introduce jump into without the "file" field of the sourcemap and will probably break a few things.

I run the tests with the compiler you linked, and I can only see a new failure wrt 0.6.9, and it's because of sourcemap without a file.

Here's the tests:

File `c.sol`:

```Solidity
pragma solidity ^0.6.0;
pragma experimental ABIEncoderV2;

import "./d.sol";

contract C {
  function test() public {
    D d = new D();

    d.fail();
  }
}
```

File `d.sol`:

```Solidity
pragma solidity ^0.6.0;
pragma experimental ABIEncoderV2;

contract D {
  struct MyStruct {
    uint x;
    uint y;
  }

  function fail() public returns (MyStruct[] memory)
  {
    revert("D failed");

    MyStruct[] memory structs = new MyStruct[](2);
    structs[0] = MyStruct(1, 2);
    structs[1] = MyStruct(2, 3);

    return structs;
  }
}
```

Deploy C, and call its `test` function.

**Comment by iamdefinitelyahuman** [comment]
No issues on my end.  In the process of testing this I did discover an issue introduced by `0.6.9`, but it's unrelated to these changes and I've already patched it.  Thanks for the heads up!

**Comment by chriseth** [comment]
@alcuadrado can you be more specific about "source maps without a file"? If the opcode is generated from an internal compiler routine used from multiple places (abi decoder, memory to storage copy routines, ...) the source file was always specified as `-1`. The change introduced in 0.6.9 was that in some circumstances, the source reference of the solidity statement that triggered the generation of the internal routine was used instead.

**Comment by alcuadrado** [comment]
By "source maps without a file" I meant "source maps with file `-1`".

I'll give more info about that failing test later today, including the exact source map that makes Buidler fail.

File: libyul/backends/evm/AsmCodeGen.cpp [hunk 1]
```diff
@@ -189,6 +189,24 @@ EthAssemblyAdapter::LabelID EthAssemblyAdapter::assemblyTagToIdentifier(evmasm::
...
 }
 
+void EthAssemblyAdapter::appendJumpInstruction(evmasm::Instruction _instruction, JumpType _jumpType)
+{
+	yulAssert(_instruction == evmasm::Instruction::JUMP || _instruction == evmasm::Instruction::JUMPI, "");
+	evmasm::AssemblyItem jump(_instruction);
+	switch (_jumpType)
+	{
+	case JumpType::Ordinary:
+		break;
```
Comment (cameel) [inline libyul/backends/evm/AsmCodeGen.cpp]:
```suggestion
	case JumpType::Ordinary:
		yulAssert(jump.getJumpType() == evmasm::AssemblyItem::JumpType::Ordinary, "");
		break;
```

**Comment by haltman-at** [comment]
OK, finally had a chance to look at this.  Answer: I don't think this should cause any serious problems for us.  (Which makes sense, as I'm the one who requested it. :) )

Such jumps will cause a bunch of `unknown function` frames to appear in the stacktraces we print out, but that's not, like, a serious incompatibility, and of course I can go ahead and modify the stacktrace code to account for this.

Haven't actually tested this out yet, this is just from looking over our code.  But will try it out as well to be sure!

(Thank you again for this!)

**Comment by chriseth** [comment]
@alcuadrado do you have any more infos for us in the meantime?

**Comment by haltman-at** [comment]
@chriseth Hey, sorry, I've been running into trouble trying to test this -- the binary doesn't work for me (it seems to expect a different version of CVC4 or something?) and I've been running into problems trying to build from source (I've asked about this on Gitter).  Could you perhaps provide a soljson I could perhaps use?  Thank you!

**Comment by chriseth** [comment]
@haltman-at the soljson is also linked in the same comment: https://365130-40892817-gh.circle-artifacts.com/0/soljson.js

**Comment by haltman-at** [comment]
Oh silly me!  Missed that somehow.  Thanks, will try that!

**Comment by haltman-at** [comment]
OK, finally tried this out and it works great!  Sorry that took so long.  Like I said, it does add some less-than-informative lines to our stacktraces, but A. I'll fix that, and B. that strikes me as basically acceptable anyway.  Thanks so much!  Now Truffle Debugger's tracking of assembly variables won't get thrown off by recursive functions. :)

**Comment by chriseth** [comment]
@alcuadrado ping

**Comment by yann300** [comment]
Just checked and it shouldn't break anything from our side

**Comment by chriseth** [comment]
Rebased.

**Comment by chriseth** [comment]
We will probably have a release this week and it would be good to get this PR in.

**Comment by alcuadrado** [comment]
Sorry for the delay @chriseth. 

Here's more data about this failing test.

The problem appears when I run a test that uses the files I pasted in [my previous comment](https://github.com/ethereum/solidity/pull/9243#issuecomment-647792524).

If you deploy the contract `C` and then call its test function, this is the execution test you get:

```
Call trace
 calling contract: d.sol:D
 value: 0
 calldata: 0xa9cc4718
 error: revert
 returnData: 0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000844206661696c6564000000000000000000000000000000000000000000000000
 steps:

    000   PUSH1 0x80                               d.sol   -  59:287
    002   PUSH1 0x40                               d.sol   -  59:287
    004   MSTORE                                   d.sol   -  59:287
    005   CALLVALUE                                d.sol   -  59:287
    006   DUP1                                     d.sol   -  59:287
    007   ISZERO                                   d.sol   -  59:287
    008   PUSH2 0x0010                             d.sol   -  59:287
    011   JUMPI (INTERNAL_JUMP)            d.sol   -  59:287
    016   JUMPDEST                                 d.sol   -  59:287
    017   POP                                      d.sol   -  59:287
    018   PUSH1 0x04                               d.sol   -  59:287
    020   CALLDATASIZE                             d.sol   -  59:287
    021   LT                                       d.sol   -  59:287
    022   PUSH2 0x002b                             d.sol   -  59:287
    025   JUMPI (INTERNAL_JUMP)            d.sol   -  59:287
    026   PUSH1 0x00                               d.sol   -  59:287
    028   CALLDATALOAD                             d.sol   -  59:287
    029   PUSH1 0xe0                               d.sol   -  59:287
    031   SHR                                      d.sol   -  59:287
    032   DUP1                                     d.sol   -  59:287
    033   PUSH4 0xa9cc4718                         d.sol   -  59:287
    038   EQ                                       d.sol   -  59:287
    039   PUSH2 0x0030                             d.sol   -  59:287
    042   JUMPI (INTERNAL_JUMP)            d.sol   -  59:287
    048   JUMPDEST                                 d.sol:D:fail   -  123:221
    049   PUSH2 0x0038                             d.sol:D:fail   -  123:221
    052   PUSH2 0x004e                             d.sol:D:fail   -  123:221
    055   JUMP (INTO_FUNCTION)             d.sol:D:fail   -  123:221
    078   JUMPDEST                                 d.sol:D:fail   -  123:221
    079   PUSH1 0x60                               d.sol:D:fail   -  155:17
    081   PUSH1 0x40                               d.sol:D:fail   -  182:18
    083   MLOAD                                    d.sol:D:fail   -  182:18
    084   PUSH32 0x08c379a000000000000000000000000000000000000000000000000000000000 d.sol:D:fail   -  182:18
    117   DUP2                                     d.sol:D:fail   -  182:18
    118   MSTORE                                   d.sol:D:fail   -  182:18
    119   PUSH1 0x04                               d.sol:D:fail   -  182:18
    121   ADD                                      d.sol:D:fail   -  182:18
    122   PUSH2 0x0082                             d.sol:D:fail   -  182:18
    125   SWAP1                                    d.sol:D:fail   -  182:18
    126   PUSH2 0x023d                             d.sol:D:fail   -  182:18
    129   JUMP (INTERNAL_JUMP)             d.sol:D:fail   -  182:18
    573   JUMPDEST                                 
    574   PUSH1 0x00                               
    576   PUSH1 0x20                               
    578   DUP3                                     
    579   ADD                                      
    580   SWAP1                                    
    581   POP                                      
    582   DUP2                                     
    583   DUP2                                     
    584   SUB                                      
    585   PUSH1 0x00                               
    587   DUP4                                     
    588   ADD                                      
    589   MSTORE                                   
    590   PUSH2 0x0256                             
    593   DUP2                                     
    594   PUSH2 0x019d                             
    597   JUMP (INTO_FUNCTION)             <------ **THIS CRASHES BUIDLER'S SOLIDITY TRACER**
    413   JUMPDEST                                 
    414   PUSH1 0x00                               
    416   PUSH2 0x01aa                             
    419   PUSH1 0x08                               
    421   DUP4                                     
    422   PUSH2 0x0296                             
    425   JUMP (INTO_FUNCTION)             <------ **THIS CRASHES BUIDLER'S SOLIDITY TRACER**
    662   JUMPDEST                                 
    663   PUSH1 0x00                               
    665   DUP3                                     
    666   DUP3                                     
    667   MSTORE                                   
    668   PUSH1 0x20                               
    670   DUP3                                     
    671   ADD                                      
    672   SWAP1                                    
    673   POP                                      
    674   SWAP3                                    
    675   SWAP2                                    
    676   POP                                      
    677   POP                                      
    678   JUMP (OUTOF_FUNCTION)            
    426   JUMPDEST                                 
    427   SWAP2                                    
    428   POP                                      
    429   PUSH32 0x44206661696c6564000000000000000000000000000000000000000000000000 
    462   PUSH1 0x00                               
    464   DUP4                                     
    465   ADD                                      
    466   MSTORE                                   
    467   PUSH1 0x20                               
    469   DUP3                                     
    470   ADD                                      
    471   SWAP1                                    
    472   POP                                      
    473   SWAP2                                    
    474   SWAP1                                    
    475   POP                                      
    476   JUMP (OUTOF_FUNCTION)            
    598   JUMPDEST                                 
    599   SWAP1                                    
    600   POP                                      
    601   SWAP2                                    
    602   SWAP1                                    
    603   POP                                      
    604   JUMP (OUTOF_FUNCTION)            
    130   JUMPDEST                                 d.sol:D:fail   -  182:18
    131   PUSH1 0x40                               d.sol:D:fail   -  182:18
    133   MLOAD                                    d.sol:D:fail   -  182:18
    134   DUP1                                     d.sol:D:fail   -  182:18
    135   SWAP2                                    d.sol:D:fail   -  182:18
    136   SUB                                      d.sol:D:fail   -  182:18
    137   SWAP1                                    d.sol:D:fail   -  182:18
    138   REVERT                                   d.sol:D:fail   -  182:18
```

The instructions that don't have a file and a range have file `-1` in their source map. Internally, in Buidler, we consider those to be unmapped.

The problem is that jumps marked as a jump into a function are unmapped. I guess this are yul functions calling other yul functions, but I don't know how these should be handled. How can/should we show that in our stack traces?

For example, if a yul function (1) calls another yul function (2) and the second one reverts, our stack trace would look something like this:

```
Revert blah blah
    ???? <-- yul function 2 reverting
    ???? <-- yul function 1 calling another yul function
    C#f() (c.sol:123)
```

The "????"s are there to represent the things that I don't know how to handle.

**Comment by haltman-at** [comment]
@alcuadrado Oh, huh, now that you bring up that sort of thing, I'm realizing I'm actually going to have to retest with Truffle as well -- I basically only tested the case I requested these for, I didn't really do a full test.  I didn't think about that sort of issue, of jumps in unmapped code.  Going to check again and see how bad this messes things up.  Oops.

**Comment by chriseth** [comment]
For some of these cases, we will have actual yul code to map to (this is done in https://github.com/ethereum/solidity/pull/9053 ) at some point, but that of course does not solve the general question.

What I can guarantee (apart from bugs) is that a "jump into" is always matched by a "jump out of", so if you do `n` jumps "into" followed by `n` jumps "out of" you are again in the same function (may it be a yul or a solidity function).

I think you can pretty safely identify functions by their entry point. So whenever you have a "jump into" that jumps to PC `x`, you know that the function that starts at PC `x` is called. You can display that more nicely to the user by checking in the source maps where that PC comes from and determine the name of the function through the AST (I guess this is what you are already doing). If you do not have a source location for that PC, you display something like `internal@<PC>`.

Is that what you were asking about, @alcuadrado ?

**Comment by chriseth** [comment]
Oh and actually another open pull request might also be relevant here: https://github.com/ethereum/solidity/pull/8868 outputs the starting PC of all functions together with the names of the functions. If you want, I can also store the names of the yul functions there...

**Comment by alcuadrado** [comment]
> For some of these cases, we will have actual yul code to map to (this is done in #9053 ) at some point, but that of course does not solve the general question.
> 
> What I can guarantee (apart from bugs) is that a "jump into" is always matched by a "jump out of", so if you do `n` jumps "into" followed by `n` jumps "out of" you are again in the same function (may it be a yul or a solidity function).

Do you mean that there are no reverts inside yul functions? Or is this excluding reverts and other terminations?


> I think you can pretty safely identify functions by their entry point. So whenever you have a "jump into" that jumps to PC `x`, you know that the function that starts at PC `x` is called. You can display that more nicely to the user by checking in the source maps where that PC comes from and determine the name of the function through the AST (I guess this is what you are already doing). 

Right now we check the source map of the jump and use that get its function using the AST.

> If you do not have a source location for that PC, you display something like `internal@<PC>`.
> Is that what you were asking about, @alcuadrado ?

Yes. We'll do this `internal@<PC>`. At least for now. It would be great if we could give more info, like `internal#enconde-struct` (not a real example, of course). Is #8868 a movement in that direction?

Thanks for your answers, @chriseth. I think this PR is a movement in the right direction. Having internal functions being treated as actual functions make everything simpler, despite needing a few changes now.

**Comment by chriseth** [comment]
> Do you mean that there are no reverts inside yul functions? Or is this excluding reverts and other terminations?

I was only talking about regular control-flow inside the currently executing "EVM frame". Opcodes like `call`, `stop`, `return` and `revert` are of course different.

> Yes. We'll do this `internal@<PC>`. At least for now. It would be great if we could give more info, like `internal#enconde-struct` (not a real example, of course). Is #8868 a movement in that direction?

Unfortunately, the internal yul functions usually have horrible names. The reason is that most of these functions are ABI coding functions and those functions always contain a description of the type and these are extremely verbose. We have to make sure there is never a collision between two such functions. A shorter example is `abi_encode_t_uint256_to_t_uint256_fromStack`. But yes, #8868 could provide these names even without the full source code that would be provided by #9053.

> Thanks for your answers, @chriseth. I think this PR is a movement in the right direction. Having internal functions being treated as actual functions make everything simpler, despite needing a few changes now.

Ah, that is very good to hear!

It would also be great to get everyone's input on #9053 and #8868 !

**Comment by alcuadrado** [comment]
> Unfortunately, the internal yul functions usually have horrible names. The reason is that most of these functions are ABI coding functions and those functions always contain a description of the type and these are extremely verbose. We have to make sure there is never a collision between two such functions. A shorter example is `abi_encode_t_uint256_to_t_uint256_fromStack`

I guess that if we get access to those names we can classify them and show that instead. For example, instead of `abi_encode_t_uint256_to_t_uint256_fromStack` show `abiEncoding`.

> It would also be great to get everyone's input on #9053 and #8868 !

Will read them tomorrow 👌

**Comment by haltman-at** [comment]
OK, on retesting with @alcuadrado's example (thanks for that by the way!), I am also seeing issues in Truffle.

Truffle doesn't crash on that jump in marking, though it does display some inelegant messages, which I should fix.  However, I consider that acceptable.

Unfortunately there is another issue: Markings in autogenerated code aren't balanced.  When I debug this transaction, what I find is that, within the autogenerated Yul for the `revert`, there are two jump in markings, but *three* jump out markings.  This messes up the final stacktrace, as it causes an extra stack frame to get popped off.

It looks to me like the final jump out marking that causes the issue is the return from autogenerated Yul back to Solidity.  If that is indeed what's causing the problem, then the obvious way to fix it would be to have the jumps *into* autogenerated Yul from Solidity be marked with a jump in to match.  And indeed it looks like the jump in was not so marked.

So, hopefully changing that will solve the problem?  Otherwise there are always blunter solutions...

**Comment by haltman-at** [comment]
Btw, worth noting that unbalanced jump markings like this causes more issues than just bad stacktraces.  I tried running our usual debugger test suite with this build, and I noticed that our test of decoding array slices failed.  On investigation, it turned out to be due to the unbalanced jump markings -- the unbalanced jump markings (since array slicing involves autogenerated Yul) confused the debugger as to which stackframe it was in, and so caused it to go looking for variables in the wrong places on the stack.

**Comment by haltman-at** [comment]
Oh hm -- it's also possible for jumps to occur in unmapped Yul code at the beginning of a transaction before the initial jump into the function being called.  That messes with some of our stuff, but in a minor way, and that I'll just consider something I should fix.  So that's not a big deal.  The unbalancing though is a real problem.

**Comment by chriseth** [comment]
Thanks for the find about the unbalanced jumps, that certainly is a real issue. I'll update this and provide you with new binaries once it is fixed.

Just a heads up: I just saw that we used "jump into" and "jump out of" already for some internal routines that are not yul functions but also not solidity functions - so this functionality is there for even older compiler versions.

**Comment by chriseth** [comment]
here is the soljson.js binary: https://371368-40892817-gh.circle-artifacts.com/0/soljson.js

**Comment by haltman-at** [comment]
> Just a heads up: I just saw that we used "jump into" and "jump out of" already for some internal routines that are not yul functions but also not solidity functions - so this functionality is there for even older compiler versions.

Interesting, I hadn't encountered that.  Can you give an example?  Is there any potential unbalance problem there?

**Comment by chriseth** [comment]
@haltman-at I can maybe supply a code example tomorrow (or rather next week, sorry), but the following internal routines exist:
 - copy array to storage
 - clear storage array
 - resize dynamic array
 - convert recursive array storage to memory
 - the "always fails assertion" default value for internal function pointers
 - the memory array and struct initialization function
 - the function that copies a contract's code to memory

note that these are inlined by the optimizer unless they are called multiple times, so you need to use the same function on the same type from multiple places to avoid inlining.

I checked and there should be no imbalance.

**Comment by haltman-at** [comment]
I see!  OK, so I can construct examples based on those if I want to test them out -- indeed I just did one and it sems to work fine; no unbalance.  So, new version seems to be fine!  Array slice test now passes.

I'll want to make the small adjustments to Truffle Debugger I mentioned, but it's not going to crash or do anything crazy or disastrous.

Thanks again! :)

**Comment by chriseth** [comment]
I think this is ready for a last round of review.

**Review by ekpyron — APPROVED** [review]
Looks good to me.

**Pull request merged**